src/main/java/com/zendesk/maxwell/recovery/Recovery.java

```diff
@@ -0,0 +1,111 @@
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import snaq.db.ConnectionPool;
+
+public class Recovery {
+	static final Logger LOGGER = LoggerFactory.getLogger(Recovery.class);
+
+	private final ConnectionPool replicationConnectionPool;
+	private final RecoveryInfo recoveryInfo;
+	private final MaxwellMysqlConfig replicationConfig;
+	private final String maxwellDatabaseName;
+	private final RecoverySchemaStore schemaStore;
+
+	public Recovery(MaxwellMysqlConfig replicationConfig,
+					String maxwellDatabaseName,
+					ConnectionPool replicationConnectionPool,
+					CaseSensitivity caseSensitivity,
+					RecoveryInfo recoveryInfo) {
+		this.replicationConfig = replicationConfig;
+		this.replicationConnectionPool = replicationConnectionPool;
+		this.recoveryInfo = recoveryInfo;
+		this.schemaStore = new RecoverySchemaStore(replicationConnectionPool, maxwellDatabaseName, caseSensitivity);
+		this.maxwellDatabaseName = maxwellDatabaseName;
+	}
+
+	public BinlogPosition recover() throws Exception {
+		String recoveryMsg = String.format(
+			"old-server-id: %d, file: %s, position: %d, heartbeat: %d",
+			recoveryInfo.serverID,
+			recoveryInfo.position.getFile(),
+			recoveryInfo.position.getOffset(),
+			recoveryInfo.heartbeat
+		);
+
+		LOGGER.warn("attempting to recover from master-change: " + recoveryMsg);
+
+		List<BinlogPosition> list = getBinlogInfo();
+		for ( int i = list.size() - 1; i >= 0 ; i-- ) {
+			BinlogPosition position = list.get(i);
+
+			LOGGER.debug("scanning binlog: " + position);
+
+			MaxwellReplicator replicator = new MaxwellReplicator(
+				this.schemaStore,
+				null,
+				null,
+				replicationConfig,
+				0L, // server-id of 0 activatives "mysqlbinlog" behavior where the server will stop after each binlog
+				false,
+				null,
+				maxwellDatabaseName,
+				position,
+				true
+			);
+
+			replicator.setFilter(new RecoveryFilter(this.maxwellDatabaseName));
+
+			BinlogPosition p = findHeartbeat(replicator);
+			if ( p != null ) {
+				LOGGER.warn("recovered new master position: " + p);
+				return p;
+			}
+		}
+
+		LOGGER.error("Could not recover from master-change: " + recoveryMsg);
+		return null;
+	}
+
+	/**
+	 * try to find a given heartbeat value from the replicator.
+	 * @return A BinlogPosition where the heartbeat was found, or null if none was found.
+	 */
+	private BinlogPosition findHeartbeat(MaxwellReplicator r) throws Exception {
+		r.startReplicator();
+		for (RowMap row = r.getRow(); row != null ; row = r.getRow()) {
+			if (Objects.equals(r.getLastHeartbeatRead(), recoveryInfo.heartbeat))
+				return row.getPosition();
+
+		}
+		return null;
+	}
+
+	/**
+	 * fetch a list of binlog postiions representing the start of each binlog file
+	 *
+	 * @return a list of binlog positions to attempt recovery at
+	 * */
+
+	private List<BinlogPosition> getBinlogInfo() throws SQLException {
```

Comment:
hm I guess you'd need to find the position associated to the last heartbeat read, and not the one we're using here (associated to last heartbeat sent) from what I understood